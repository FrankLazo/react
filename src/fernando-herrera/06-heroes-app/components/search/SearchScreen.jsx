import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import useForm from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import getHeroesByName from '../../selectors/getHeroesByName';
import { useMemo } from 'react';

const SearchScreen = ({ history }) =>
{
	const { search } = useLocation();
	const { q = '' } = queryString.parse(search);

	const { values, handleInputChange } = useForm({ searchText: q, });
	const { searchText } = values;

	const filteredHeroes = useMemo(() => getHeroesByName(q), [q]);

	const handleSubmit = (e) =>
	{
		e.preventDefault();
		history.push(`?q=${ searchText }`);
	};

	return (
		<div className="content m-4">
			<h1 className="has-text-primary">Search</h1>
			<hr className="has-background-grey-lighter" />

			<div className="columns">
				<div className="column is-one-third">
					<h2 className="is-size-4">Search form</h2>

					<form className="form" onSubmit={ handleSubmit }>
						<div className="field">
							<div className="control">
								<input
									type="text"
									name="searchText"
									value={ searchText }
									className="input"
									placeholder="Find your hero"
									onChange={ handleInputChange }
									/>
							</div>
						</div>

						<div className="field">
							<div className="control">
								<button
									type="submit"
									className="button is-primary is-outlined is-flex full-width"
									>
									Search
								</button>
							</div>
						</div>
					</form>
				</div>

				<div className="column">
					<h2 className="is-size-4">Results</h2>

					{
						(q === '') && (
							<div className="notification is-warning is-light">Search a hero...</div>
						)
					}

					{
						(q !== '' && filteredHeroes.length === 0) && (
							<div className="notification is-danger is-light">
								There is no a hero with "{ q }"
							</div>
						)
					}

					<div className="columns is-multiline">
					{
						filteredHeroes.map(hero => (
							<HeroCard key={ hero.id } { ...hero } />
						))
					}
					</div>
				</div>
			</div>
		</div>
	);
};

export default SearchScreen;
