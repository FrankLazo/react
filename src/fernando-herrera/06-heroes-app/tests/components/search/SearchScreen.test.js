import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

describe('<SearchScreen /> tests', () => {
	const historyMock = {
		push: jest.fn(),
	};

	test('should show component with default values', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<Route path="/search" component={ SearchScreen } />
			</MemoryRouter>
		);

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('button').text().trim()).toBe('Search');
	});

	test('should show component with queryString value', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<Route path="/search" component={ SearchScreen } />
			</MemoryRouter>
		);

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('input').prop('value').trim()).toBe('batman');
	});

	test('should show error if hero is not found', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman123']}>
				<Route path="/search" component={ SearchScreen } />
			</MemoryRouter>
		);

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('.is-danger').exists()).toBe(true);
	});

	test('should call to history\'s push', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<Route
					path="/search"
					component={() => <SearchScreen history={ historyMock } /> } />
			</MemoryRouter>
		);

		wrapper.find('input').simulate('change', {
			target: {
				name: 'searchText',
				value: 'batman',
			},
		});

		const handleSubmit = wrapper.find('form').prop('onSubmit');

		handleSubmit({ preventDefault(){} });
		expect(historyMock.push).toHaveBeenCalledWith('?q=batman');
	});
});
