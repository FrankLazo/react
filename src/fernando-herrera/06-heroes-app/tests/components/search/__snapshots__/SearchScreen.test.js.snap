// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> tests should show component with default values 1`] = `
<div
  className="content m-4"
>
  <h1
    className="has-text-primary"
  >
    Search
  </h1>
  <hr
    className="has-background-grey-lighter"
  />
  <div
    className="columns"
  >
    <div
      className="column is-one-third"
    >
      <h2
        className="is-size-4"
      >
        Search form
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <div
            className="control"
          >
            <input
              className="input"
              name="searchText"
              onChange={[Function]}
              placeholder="Find your hero"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="control"
          >
            <button
              className="button is-primary is-outlined is-flex full-width"
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="column"
    >
      <h2
        className="is-size-4"
      >
        Results
      </h2>
      <div
        className="notification is-warning is-light"
      >
        Search a hero...
      </div>
      <div
        className="columns is-multiline"
      />
    </div>
  </div>
</div>
`;

exports[`<SearchScreen /> tests should show component with queryString value 1`] = `
<div
  className="content m-4"
>
  <h1
    className="has-text-primary"
  >
    Search
  </h1>
  <hr
    className="has-background-grey-lighter"
  />
  <div
    className="columns"
  >
    <div
      className="column is-one-third"
    >
      <h2
        className="is-size-4"
      >
        Search form
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <div
            className="control"
          >
            <input
              className="input"
              name="searchText"
              onChange={[Function]}
              placeholder="Find your hero"
              type="text"
              value="batman"
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="control"
          >
            <button
              className="button is-primary is-outlined is-flex full-width"
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="column"
    >
      <h2
        className="is-size-4"
      >
        Results
      </h2>
      <div
        className="columns is-multiline"
      >
        <div
          className="column is-half-tablet is-one-third-desktop is-one-quarter-widescreen"
        >
          <div
            className="card"
          >
            <div
              className="card-image"
            >
              <img
                alt="Batman"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="card-content"
            >
              <div
                className="media"
              >
                <div
                  className="media-content"
                >
                  <p
                    className="title is-4"
                  >
                    Batman
                  </p>
                  <p
                    className="subtitle is-6"
                  >
                    Bruce Wayne
                  </p>
                </div>
              </div>
              <div
                className="content"
              >
                <p
                  className="has-text-grey-light"
                >
                  Detective Comics #27
                </p>
              </div>
            </div>
            <footer
              className="card-footer"
            >
              <a
                className="card-footer-item"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Ver m√°s...
              </a>
            </footer>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchScreen /> tests should show error if hero is not found 1`] = `
<div
  className="content m-4"
>
  <h1
    className="has-text-primary"
  >
    Search
  </h1>
  <hr
    className="has-background-grey-lighter"
  />
  <div
    className="columns"
  >
    <div
      className="column is-one-third"
    >
      <h2
        className="is-size-4"
      >
        Search form
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <div
            className="control"
          >
            <input
              className="input"
              name="searchText"
              onChange={[Function]}
              placeholder="Find your hero"
              type="text"
              value="batman123"
            />
          </div>
        </div>
        <div
          className="field"
        >
          <div
            className="control"
          >
            <button
              className="button is-primary is-outlined is-flex full-width"
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="column"
    >
      <h2
        className="is-size-4"
      >
        Results
      </h2>
      <div
        className="notification is-danger is-light"
      >
        There is no a hero with "
        batman123
        "
      </div>
      <div
        className="columns is-multiline"
      />
    </div>
  </div>
</div>
`;
